File: environment/src/test/java/jetbrains/exodus/tree/btree/BTreeCursorDeleteTest.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.btree;

import jetbrains.exodus.bindings.IntegerBinding;
import jetbrains.exodus.env.Cursor;
import org.junit.Assert;
import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.*;

public class BTreeCursorDeleteTest extends BTreeTestBase {

    @Test
    public void testDeleteCursorNoDuplicates1() throws IOException {
        tm = createMutableTree(false, 1);
        getTreeMutable().put(kv(1, "1"));

        Cursor c = tm.openCursor();

        assertTrue(c.getNext());
        assertTrue(c.deleteCurrent());
        assertFalse(c.deleteCurrent());
        assertFalse(c.getNext());
    }

    @Test
    public void testDeleteCursorNoDuplicates2() throws IOException {
        tm = createEmptyTreeForCursor(1).getMutableCopy();

        for (int i = 0; i < 8; i++) {
            getTreeMutable().put(kv(i, "v" + i));
        }

        Cursor c = tm.openCursor();
        assertTrue(c.getNext());
        assertEquals(key(0), c.getKey());
        assertTrue(c.deleteCurrent());
        assertFalse(c.deleteCurrent());

        assertTrue(c.getNext());
        assertEquals(key(1), c.getKey());

        assertEquals(value("v7"), c.getSearchKey(key(7)));
        assertTrue(c.deleteCurrent());
        assertFalse(c.deleteCurrent());
        assertFalse(c.getNext());
    }

    @Test
    public void testDeleteCursorDuplicates1() throws IOException {
        tm = createMutableTree(true, 1);
        getTreeMutable().put(kv(1, "11"));
        getTreeMutable().put(kv(1, "12"));

        assertTrue(getTreeMutable().getRoot() instanceof BottomPageMutable);
        assertEquals(1, getTreeMutable().getRoot().getSize());
        assertTrue(getTreeMutable().getRoot().getKey(0) instanceof LeafNodeDupMutable);
        assertEquals(2, getTreeMutable().getRoot().getKey(0).getDupCount());

        Cursor c = getTreeMutable().openCursor();

        assertTrue(c.getNext());
        assertTrue(c.deleteCurrent());
        assertFalse(c.deleteCurrent());

        assertTrue(getTreeMutable().getRoot() instanceof BottomPageMutable);
        assertEquals(1, getTreeMutable().getRoot().getSize());
        assertTrue(getTreeMutable().getRoot().getKey(0) instanceof LeafNodeMutable);

        assertTrue(c.getNext());
        assertTrue(c.deleteCurrent());
        assertFalse(c.deleteCurrent());
        assertFalse(c.getNext());
    }

    @Test
    public void testDeleteCursorDuplicates2() throws IOException {
        tm = createEmptyTreeForCursor(1).getMutableCopy();

        for (int i = 0; i < 8; i++) {
            getTreeMutable().put(kv(i, "v" + i));
            getTreeMutable().put(kv(i, "vv" + i));
        }

        Cursor c = getTreeMutable().openCursor();

        assertEquals(true, c.getSearchBoth(key(1), value("vv1")));
        assertTrue(c.deleteCurrent());
        assertFalse(c.deleteCurrent());
        assertTrue(c.getNext());

        assertEquals(true, c.getSearchBoth(key(7), value("vv7")));
        assertTrue(c.deleteCurrent());
        assertFalse(c.getNext());
    }

    @Test
    public void testDeleteCursorDuplicates3() throws IOException {
        tm = createMutableTree(true, 1).getMutableCopy();

        for (int i = 0; i < 32; ++i) {
            for (int j = 0; j < 32; ++j) {
                getTreeMutable().put(IntegerBinding.intToEntry(i), IntegerBinding.intToEntry(j));
            }
        }

        for (int i = 0; i < 32; ++i) {
            final Cursor cursor = getTreeMutable().openCursor();
            Assert.assertNotNull(cursor.getSearchKeyRange(IntegerBinding.intToEntry(i)));
            for (int j = 0; j < 31; ++j) {
                cursor.deleteCurrent();
                Assert.assertTrue(cursor.getNext());
            }
        }

    }

}

File: environment/src/test/kotlin/jetbrains/exodus/tree/btree/BTreeCursorDeleteTest.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.btree

import jetbrains.exodus.bindings.IntegerBinding
import jetbrains.exodus.env.Cursor
import org.junit.Assert
import org.junit.Test

class BTreeCursorDeleteTest : BTreeTestBase() {
    @Test
    fun testDeleteCursorNoDuplicates1() {
        treeMutable = createMutableTree(false, 1)
        treeMutable!!.put(kv(1, "1"))
        val c: Cursor = treeMutable!!.openCursor()
        Assert.assertTrue(c.next)
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.deleteCurrent())
        Assert.assertFalse(c.next)
    }

    @Test
    fun testDeleteCursorNoDuplicates2() {
        treeMutable = createEmptyTreeForCursor(1).mutableCopy
        for (i in 0..7) {
            treeMutable!!.put(kv(i, "v$i"))
        }
        val c: Cursor = treeMutable!!.openCursor()
        Assert.assertTrue(c.next)
        Assert.assertEquals(key(0), c.key)
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.deleteCurrent())
        Assert.assertTrue(c.next)
        Assert.assertEquals(key(1), c.key)
        Assert.assertEquals(value("v7"), c.getSearchKey(key(7)))
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.deleteCurrent())
        Assert.assertFalse(c.next)
    }

    @Test
    fun testDeleteCursorDuplicates1() {
        treeMutable = createMutableTree(true, 1)
        treeMutable!!.put(kv(1, "11"))
        treeMutable!!.put(kv(1, "12"))
        Assert.assertTrue(treeMutable!!.root is BottomPageMutable)
        Assert.assertEquals(1, treeMutable!!.root.size.toLong())
        Assert.assertTrue(treeMutable!!.root.getKey(0) is LeafNodeDupMutable)
        Assert.assertEquals(2, treeMutable!!.root.getKey(0).dupCount)
        val c: Cursor = treeMutable!!.openCursor()
        Assert.assertTrue(c.next)
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.deleteCurrent())
        Assert.assertTrue(treeMutable!!.root is BottomPageMutable)
        Assert.assertEquals(1, treeMutable!!.root.size.toLong())
        Assert.assertTrue(treeMutable!!.root.getKey(0) is LeafNodeMutable)
        Assert.assertTrue(c.next)
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.deleteCurrent())
        Assert.assertFalse(c.next)
    }

    @Test
    fun testDeleteCursorDuplicates2() {
        treeMutable = createEmptyTreeForCursor(1).mutableCopy
        for (i in 0..7) {
            treeMutable!!.put(kv(i, "v$i"))
            treeMutable!!.put(kv(i, "vv$i"))
        }
        val c: Cursor = treeMutable!!.openCursor()
        Assert.assertTrue(c.getSearchBoth(key(1), value("vv1")))
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.deleteCurrent())
        Assert.assertTrue(c.next)
        Assert.assertTrue(c.getSearchBoth(key(7), value("vv7")))
        Assert.assertTrue(c.deleteCurrent())
        Assert.assertFalse(c.next)
    }

    @Test
    fun testDeleteCursorDuplicates3() {
        treeMutable = createMutableTree(true, 1)!!.mutableCopy
        for (i in 0..31) {
            for (j in 0..31) {
                treeMutable!!.put(IntegerBinding.intToEntry(i), IntegerBinding.intToEntry(j))
            }
        }
        for (i in 0..31) {
            val cursor: Cursor = treeMutable!!.openCursor()
            Assert.assertNotNull(cursor.getSearchKeyRange(IntegerBinding.intToEntry(i)))
            for (j in 0..30) {
                cursor.deleteCurrent()
                Assert.assertTrue(cursor.next)
            }
        }
    }
}
